### PR TITLE
feat(confidential/builder): add ordered confidential Batch assembler

### PR DESCRIPTION
**Pull request by aluque-peersyst**
# feat(confidential/builder): add ordered confidential Batch assembler

## Description
This PR aims to add `BuildBatch` to `confidential/builder`, which assembles two to eight ordered confidential MPT operations into one XLS-56 `Batch`. Calling the standalone builders in a row cannot produce a valid `Batch`: each reads the ledger, and inside a `Batch` the ledger does not yet reflect what an earlier inner leaves behind, so every proof after the first binds a balance and version the transaction will not find when it applies. `BuildBatch` reads every `MPToken` and `MPTokenIssuance` from one validated ledger and threads predicted state through the inners in order.

Depends on the GetSpendingBalance PR; this branch is stacked on it.

## Type of change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Documentation update
- [ ] Refactoring

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where needed
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes

## Changes

- Add `BuildBatch`, `BuildBatchParams`, and the operation types `ConvertOp`, `ConvertBackOp`, `SendOp`, `MergeInboxOp`, `ClawbackOp`, and `TransactionOp`.
- Predict per-inner state exactly as the xrpld transactors apply it: spending and inbox ciphertexts, issuer and auditor mirrors, holder keys, balance versions, public amounts, and the send re-randomization of destination credits.
- Resolve every inner nonce before generating proofs. Same-account inners take consecutive sequences starting one past the outer `Batch`; a `TransactionOp` may carry its own `Sequence` only if it matches its position (`ErrBatchInnerSequenceMismatch`) or a `TicketSequence`.
- Shape inners for XLS-56 with `tfInnerBatchTxn`, zero `Fee`, empty `SigningPubKey`, and no signature. Only `tfAllOrNothing` is supported.
- Refuse chains that read a balance an earlier merge, clawback, or first-time convert reset to the canonical encrypted zero (`ErrBatchUnpredictableState`), and ready-made inners outside an allowlist of types that cannot touch confidential state (`ErrBatchInnerNotSupported`).
- Add `elgamal.Add` and `elgamal.Subtract` over new `mptcrypto.AddCiphertexts` and `mptcrypto.SubtractCiphertexts` bindings.
- Add unit tests, a localnet integration test for a dependent batch, a builders guide section, and a changelog entry.

### CODE REVIEW

File: confidential/builder/batch_state.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+// knownBalance is a ciphertext read from the ledger or predicted from one.
+func knownBalance(ciphertext string) predictedBalance {
+	return predictedBalance{value: ciphertext}
+}
+
+// unpredictableBalance marks a field that exists but whose value this package cannot reproduce,
+// because the transactor derived it from the canonical encrypted zero.
+func unpredictableBalance() predictedBalance {
+	return predictedBalance{unpredictable: true}
```
Comment (kpitapeersyst) [inline confidential/builder/batch_state.go]: The claim that the client cannot reproduce this ciphertext is incorrect. The bundled `confidential/deps/include/secp256k1_mpt.h` declares `generate_canonical_encrypted_zero`, and the bundled Linux `libmpt-crypto.a` exports it. xrpld's `encryptCanonicalZeroAmount` calls this same function with public inputs: the encryption key, account ID, and issuance ID.

Please bind this function through `mptcrypto` and use the actual ciphertexts when a convert, merge, or clawback resets a balance. This lets us remove the `unpredictable` flag, its propagation branches, the special error, and the artificial restrictions on otherwise valid operation chains. This is an opportunity to delete a whole category of state complexity, rather than maintain a workaround for an available primitive.

Reference: https://github.com/XRPLF/rippled/blob/develop/src/libxrpl/protocol/ConfidentialTransfer.cpp (`encryptCanonicalZeroAmount`).

File: confidential/builder/batch_operations.go [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+	return mirrors
+}
+
+// supportedInnerTransactionTypes is the set of ordinary transaction types the assembler accepts as
+// a ready-made inner.
+var supportedInnerTransactionTypes = map[transaction.TxType]struct{}{
+	transaction.AccountSetTx:       {},
+	transaction.SetRegularKeyTx:    {},
+	transaction.SignerListSetTx:    {},
+	transaction.TicketCreateTx:     {},
```
Comment (kpitapeersyst) [inline confidential/builder/batch_operations.go:356]: The allowlist checks confidential-state effects, but this transaction also changes the account sequence by `TicketCount` in addition to normal transaction sequence consumption. `buildPlainInner` only allocates or claims one sequence.

I reproduced this through `BuildBatch`: with outer sequence `10`, a `TicketCreate` with count `3` gets sequence `11`, and the following `AccountSet` gets `12`. The builder returns success, but the second inner requires sequence `15`. A following confidential operation would also bind its proof to the wrong nonce.

Please reject `TicketCreate` until its sequence effects are modeled, or include those effects in ordered nonce allocation. Add coverage for both sequence-funded and ticket-funded creation. The latter also advances the account sequence by the number of tickets created.

Reference: https://github.com/XRPLF/rippled/blob/develop/src/libxrpl/tx/transactors/system/TicketCreate.cpp (`doApply`).

File: confidential/builder/batch_state.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+		}
+		*field.target = knownBalance(ciphertext)
+	}
+	if state.version, err = optionalUint32(resp.Node, "ConfidentialBalanceVersion"); err != nil {
+		return nil, err
+	}
+	if state.publicAmount, err = optionalUint64(resp.Node, "MPTAmount"); err != nil {
+		return nil, err
+	}
+	if err := requireCurrentBalanceVersion(snapshot, resp.Index, state.version); err != nil {
```
Comment (kpitapeersyst) [inline confidential/builder/batch_state.go]: This checks the current balance version of every loaded token, including tokens used only by converts or as send destinations. The standalone `getMPTokenConvertState` deliberately does not read the version because conversion does not consume it.

For example, two repeat converts can now fail with `ErrStaleBalanceVersion` because a pending send changed the holder's spending version, even though neither convert binds that version and that send does not change the public balance being converted.

Please share operation-specific state requirements between standalone and batch builders, and require version freshness where a proof consumes that version rather than whenever a token is loaded. The duplicated preflight policy has already produced different acceptance rules.

File: confidential/builder/batch_test.go [hunk 4]
```diff
@@ -0,0 +1,1143 @@
...
+	t.Helper()
+
+	index, err := xrplhash.MPToken(issuanceID, holder)
+	require.NoError(t, err)
+	mirror, ok := fixture.entries[index]["IssuerEncryptedBalance"].(string)
+	require.True(t, ok, "fixture holder has no issuer mirror balance")
+	return mirror
+}
+
+func TestBuildBatchUnreadableBalancesStillUsable(t *testing.T) {
```
Comment (kpitapeersyst) [inline confidential/builder/batch_test.go]: This commit introduces a 1,143-line file that combines proof-binding scenarios, nonce allocation, input validation, preflight policy, and state-transition tests. Can we split it by those responsibilities first? The shared fixtures can remain in `batch_helpers_test.go`.

The behaviors do not require one large file, and focused files would make it easier to locate the tests that protect each rule. Keep the distinct scenarios and assertions, but separate their ownership.

File: confidential/builder/batch.go [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+		nonces.next[batchAccount.Classic] = p.Sequence + 1
+	}
+	for _, operation := range p.Operations {
+		ownNonce, err := operationCarriesNonce(operation)
+		if err != nil {
+			return nil, err
+		}
+		if ownNonce {
+			continue
+		}
```
Comment (kpitapeersyst) [inline confidential/builder/batch.go]: A supplied nonce currently means this operation is skipped during account tracking. If a non-outer account appears only in ready-made inners with explicit sequences, its counter is never loaded, and `batchNonces.claim` silently succeeds when that account is untracked. Tickets likewise return from `allocate` without reserving their value.

I reproduced both failures through `BuildBatch`, without any `TicketCreate`:
- With outer sequence `10` and another account currently at `5`, two distinct `AccountSet` inners from that other account with sequences `5, 7` return success. The second inner actually requires `6`. Using `5, 5` also returns success.
- Two confidential sends from the same account, with different amounts but the same `TicketSequence: 900`, return success with both proofs generated. xrpld rejects duplicate inner nonces in all-or-nothing mode.

Please resolve one ordered nonce plan for all operations before generating proofs. Keep caller-set sequences in their account's counter and track used tickets instead of treating a supplied nonce as permission to bypass validation. Reserve an outer ticket too, so an inner cannot spend it again. Cover explicit-only secondary accounts and reused tickets in regression tests. This is separate from the extra sequence consumption caused by `TicketCreate`.

Reference: https://github.com/XRPLF/rippled/blob/551a19b10d90eadd681a84c78177a8e95ad371d4/src/libxrpl/tx/transactors/system/Batch.cpp#L374-L400

File: confidential/builder/batch_operations.go [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+func buildPlainInner(nonces *batchNonces, op TransactionOp) (transaction.FlatTransaction, error) {
+	flat := op.Tx.Flatten()
+	account, err := plainInnerAccount(op)
+	if err != nil {
+		return nil, err
+	}
+	ticket, _ := flat["TicketSequence"].(uint32)
+	sequence, _ := flat["Sequence"].(uint32)
+	switch {
+	case ticket != 0:
```
Comment (kpitapeersyst) [inline confidential/builder/batch_operations.go]: The ticket branch silently takes precedence when a `TransactionOp` carries both nonces, but leaves both fields in the output. Neither `validatePreparedTransaction(op.Tx)` nor the final `Batch.Validate()` checks their mutual exclusion.

I reproduced this with outer sequence `10`, an `AccountSet` carrying `Sequence: 11` and `TicketSequence: 900`, and a second ordinary inner. `BuildBatch` returns success and preserves both non-zero fields in the first inner. xrpld requires a ticketed inner to have `Sequence: 0`, so this batch cannot be accepted.

Please extract the ready-made inner's typed account and nonce options once and apply the existing `TxOptions.validate` contract during preflight, before ledger queries or proof generation. Reuse `ErrConflictingNonce` rather than maintaining a separate, weaker nonce path for `TransactionOp`. Add a regression test that expects this error instead of a successfully built batch.

Reference: https://github.com/XRPLF/rippled/blob/551a19b10d90eadd681a84c78177a8e95ad371d4/src/libxrpl/tx/transactors/system/Batch.cpp#L353-L372

File: confidential/builder/batch.go [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+		return validatePlainInner(plain)
+	}
+	options, err := operationOptions(operation)
+	if err != nil {
+		return err
+	}
+	if options.Sequence != 0 {
+		return ErrBatchInnerSequenceSet
+	}
+	return nil
```
Comment (kpitapeersyst) [inline confidential/builder/batch.go]: Validate every operation before ledger queries or proof work

This preflight only checks the operation kind and its sequence. The canonical base validators and balance-range checks do not run until `prepareBatch*`, after sequence/state loading and potentially after earlier operations have generated proofs. That changes the public error contract as well as doing avoidable work.

I reproduced two cases against the standalone builder:
- A `SendOp` with `Destination: "not-an-address"` makes two account queries and returns `ErrInvalidAddress` from `newTokenKey`. `BuildSend` returns `ErrInvalidDestination` with no queries.
- A later `SendOp` with `Amount: 0` returns `ErrLedgerQuery` when the endpoint is unavailable. `BuildSend` rejects the same parameters locally with `ErrZeroAmount`.

Please run each operation's existing base validator, balance-range validation, and applicable `TxOptions` validation in the initial operation-validation pass. Keep these input-only rules separate from checks against predicted ledger state. This preserves the documented standalone sentinels and rejects invalid later operations before earlier ones incur query or proof costs. Add regression coverage for both the error identity and zero query calls.

File: confidential/builder/batch.go [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+	Flatten() transaction.FlatTransaction
+	Validate() (bool, error)
+}
+
+func (ConvertOp) batchOperation()     {}
+func (ConvertBackOp) batchOperation() {}
+func (SendOp) batchOperation()        {}
+func (MergeInboxOp) batchOperation()  {}
+func (ClawbackOp) batchOperation()    {}
+func (TransactionOp) batchOperation() {}
```
Comment (kpitapeersyst) [inline confidential/builder/batch.go]: Make the operation interface match the values the builder accepts

These value-receiver marker methods mean both `SendOp` and `*SendOp`, and likewise every other operation pair, satisfy `BatchOperation`. However, the downstream type switches and `TransactionOp` assertions accept only values. A non-nil pointer compiles as a supported operation and then always fails with `ErrBatchInnerNotSupported`. I reproduced this through `BuildBatch` with a pointer to an otherwise valid `TransactionOp`.

The marker-only interface also makes account extraction, nonce extraction, state references, and preparation maintain separate lists of the same operation types. Please normalize each operation once into a validated internal plan, or give the sealed interface the behavior the builder actually needs. Either approach should make supported pointer/value forms consistent and reject nil pointers explicitly. Do not fix this by duplicating pointer cases across every switch. Add a regression test for non-nil pointer operations.

**Review by kpitapeersyst — CHANGES_REQUESTED** [review]
Good work on this!

The ordered state prediction and proof-binding tests provide a solid foundation for confidential batches.

I’ve left eight comments covering nonce handling, validation, state modeling, and code structure. The main goal is to fix the correctness gaps and reuse existing helpers so we remove special cases rather than add more.

Once these points are addressed and the dependent flows pass live integration tests, this should be in good shape.

**Review by kpitapeersyst — APPROVED** [review]
